File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
        def isSymmetric(self, root: Optional[TreeNode]) -> bool:
            if not root.left and not root.right:
                return True

            return self.isMirror(root.left, root.right)

        def isMirror(self, l_node: Optional[TreeNode], r_node: Optional[TreeNode]) -> bool:
            if not l_node and not r_node:
                return True

            if not l_node or not r_node:
                return False

            if l_node.val != r_node.val:
                return False

            return self.isMirror(l_node.left, r_node.right) and self.isMirror(l_node.right, r_node.left)

File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isMirror(root.left, root.right);
    }

    private boolean isMirror(TreeNode lNode, TreeNode rNode) {
        if (lNode == null && rNode == null) {
            return true;
        }

        if (lNode == null || rNode == null) {
            return false;
        }

        if (lNode.val != rNode.val) {
            return false;
        }

        return isMirror(lNode.left, rNode.right) && isMirror(lNode.right, rNode.left);
    }
}